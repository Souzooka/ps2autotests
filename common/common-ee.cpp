#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
