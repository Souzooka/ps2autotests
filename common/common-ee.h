#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
