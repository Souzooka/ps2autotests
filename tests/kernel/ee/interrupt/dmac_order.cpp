#include <common-ee.h>
#include <kernel.h>
#include <ee_regs.h>
#include <string.h>

#define MAX_DMAC_RESULTS 9
#define MAX_DMAC_HANDLERS 10

static u32 g_dmacSema = 0;
static u32 g_dmacHandlers[MAX_DMAC_HANDLERS];
static u32 g_dmacHandlerCount = 0;
static u32 g_orderPos = 0;
static u32 g_order[MAX_DMAC_RESULTS];

static const u32 __attribute__((aligned(16))) g_dummyTransferData[4] = {0, 0, 0, 0};

typedef s32 (*DmacHandlerFunction)(s32);

void appendResult(u32 result) {
	if(g_orderPos != MAX_DMAC_RESULTS) {
		iSignalSema(g_dmacSema);
		g_order[g_orderPos] = result;
		g_orderPos++;
	}
}

template <s32 number>
s32 continueDmacHandler(s32 channel) {
	appendResult(number);
	return 0;
}

template <s32 number>
s32 cancelDmacHandler(s32 channel) {
	appendResult(number);
	return -1;
}

void initializeTest() {
	g_orderPos = 0;
	g_dmacHandlerCount = 0;
	{
		ee_sema_t dmacSemaInfo;
		memset(&dmacSemaInfo, 0, sizeof(dmacSemaInfo));
		dmacSemaInfo.init_count = 0;
		dmacSemaInfo.max_count  = MAX_DMAC_RESULTS;
		dmacSemaInfo.option     = 0;
		g_dmacSema = CreateSema(&dmacSemaInfo);
	}
}

void addTestDmacHandler(DmacHandlerFunction handler, s32 next) {
	int handlerId = AddDmacHandler(DMAC_TO_SPR, handler, next);
	g_dmacHandlers[g_dmacHandlerCount] = handlerId;
	g_dmacHandlerCount++;
}

void removeTestDmacHandler(int index) {
	RemoveDmacHandler(DMAC_TO_SPR, g_dmacHandlers[index]);
	for(u32 i = index + 1; i < g_dmacHandlerCount; i++) {
		g_dmacHandlers[i - 1] = g_dmacHandlers[i];
	}
	g_dmacHandlerCount--;
}

void waitForTestResults() {
	EnableDmac(DMAC_TO_SPR);
	for(u32 i = 0; i < MAX_DMAC_RESULTS; i++) {
		*(R_EE_D9_MADR) = (u32)g_dummyTransferData;
		*(R_EE_D9_SADR) = 0;
		*(R_EE_D9_QWC)  = 1;
		*(R_EE_D9_CHCR) = 0x100;
		while(*(R_EE_D9_CHCR) & 0x100) {}
		WaitSema(g_dmacSema);
	}
	DisableDmac(DMAC_TO_SPR);
}

void finishTest() {
	for(u32 i = 0; i < MAX_DMAC_RESULTS; i++) {
		if(i != 0) {
			printf(", %d", g_order[i]);
		} else {
			printf("%d", g_order[i]);
		}
	}
	printf("\n");
	for(u32 i = 0; i < g_dmacHandlerCount; i++) {
		RemoveDmacHandler(DMAC_TO_SPR, g_dmacHandlers[i]);
	}
}

void pushBackTest() {
	printf("push back: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, -1);
	addTestDmacHandler(&continueDmacHandler<2>, -1);
	addTestDmacHandler(&continueDmacHandler<3>, -1);
	waitForTestResults();
	finishTest();
}

void pushFrontTest() {
	printf("push front: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, 0);
	addTestDmacHandler(&continueDmacHandler<2>, 0);
	addTestDmacHandler(&continueDmacHandler<3>, 0);
	waitForTestResults();
	finishTest();
}

void mixedFrontBackTest() {
	printf("mixed front back: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, -1);
	addTestDmacHandler(&continueDmacHandler<2>,  0);
	addTestDmacHandler(&continueDmacHandler<3>, -1);
	addTestDmacHandler(&continueDmacHandler<4>,  0);
	waitForTestResults();
	finishTest();
}

void relativeTest() {
	printf("relative: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, -1);
	addTestDmacHandler(&continueDmacHandler<2>, -1);
	addTestDmacHandler(&continueDmacHandler<3>, -1);
	addTestDmacHandler(&continueDmacHandler<4>, g_dmacHandlers[1]);
	waitForTestResults();
	finishTest();
}

void removeTest() {
	printf("remove: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, -1);
	addTestDmacHandler(&continueDmacHandler<2>, -1);
	addTestDmacHandler(&continueDmacHandler<3>, -1);
	removeTestDmacHandler(1);
	waitForTestResults();
	finishTest();
}

void cancelTest() {
	printf("cancel: ");
	initializeTest();
	addTestDmacHandler(&continueDmacHandler<1>, -1);
	addTestDmacHandler(&cancelDmacHandler<2>,   -1);
	addTestDmacHandler(&continueDmacHandler<3>, -1);
	waitForTestResults();
	finishTest();
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	pushBackTest();
	pushFrontTest();
	mixedFrontBackTest();
	removeTest();
	relativeTest();
	cancelTest();
	
	printf("-- TEST END\n");
	
	return 0;
}
