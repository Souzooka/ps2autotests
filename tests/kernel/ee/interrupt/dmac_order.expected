-- TEST BEGIN
push back: 1, 2, 3, 1, 2, 3, 1, 2, 3
push front: 3, 2, 1, 3, 2, 1, 3, 2, 1
mixed front back: 4, 2, 1, 3, 4, 2, 1, 3, 4
remove: 1, 3, 1, 3, 1, 3, 1, 3, 1
relative: 1, 4, 2, 3, 1, 4, 2, 3, 1
cancel: 1, 2, 1, 2, 1, 2, 1, 2, 1
-- TEST END
